cmd/kubelet/app/server.go

```diff
@@ -412,13 +431,249 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
+	}
+
+	return nil
+}
+
+// getCurrentContextAuthInfo returns the AuthInfo object that's referenced
+// by the current context.
+// If current context or auth info name is empty, then it will return an error.
+// If AuthInfo is empty, a new auth info object will be created.
+func getCurrentContextAuthInfo(config *clientcmdapi.Config) (*clientcmdapi.AuthInfo, error) {
+	ctx, ok := config.Contexts[config.CurrentContext]
+	if !ok {
+		return nil, fmt.Errorf("unable to find current context %q", config.CurrentContext)
+	}
+
+	if ctx.AuthInfo == "" {
+		return nil, fmt.Errorf("unable to find the name of the authInfo in current context %q", config.CurrentContext)
+	}
+
+	if _, ok := config.AuthInfos[ctx.AuthInfo]; !ok {
+		config.AuthInfos[ctx.AuthInfo] = clientcmdapi.NewAuthInfo()
+	}
+	return config.AuthInfos[ctx.AuthInfo], nil
+}
+
+// TODO(yifan): More detailed check on the cert / key content.
+// CheckTLSInfo returns true if the authInfo contains client certificate data for client key data.
+// Or if the client certificate or key file exists.
+func containsSufficientTLSInfo(authInfo *clientcmdapi.AuthInfo) bool {
+	// We use '||' so that we won't override existing data in case of wrong setup.
+	if len(authInfo.ClientCertificateData) > 0 || len(authInfo.ClientKeyData) > 0 {
+		return true
+	}
+
+	if crypto.FoundCertOrKey(authInfo.ClientCertificate, authInfo.ClientKey) {
+		return true
+	}
+
+	return false
+}
+
+// getClientCertAndKey will:
+// (1) Create a restful client for doing the certificate signing request.
+// (2) Read existing key data from existingKeyPath if possible.
+// (3) Pass 'requestClientCertificate()' the CSR client, existing key data, and node name to
+//     request for client certificate from the API server.
+// (4) Once (3) succeeds, dump the certificate and key data to the given paths.
+// On failure, the the certificate and key file will be cleaned up.
+// If the existingCertPath or existingKeyPath is empty, then the function will use the default path, respectively:
+// /var/run/kubernetes/kubelet-client.crt, /var/run/kubernetes/kubelet-client.key.
+func getClientCertAndKey(s *options.KubeletServer, existingCertPath, existingKeyPath string) (certPath, keyPath string, err error) {
+	// (1).
+	clientConfig, err := kubeconfigClientConfig(s.BootstrapKubeconfig, s.APIServerList)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to create client config: %v", err)
+	}
+	client, err := unversionedcertificates.NewForConfig(clientConfig)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to create certificates signing request client: %v", err)
+	}
+	csrClient := client.CertificateSigningRequests()
+
+	// (2).
+	certPath, keyPath = existingCertPath, existingKeyPath
+	if certPath == "" {
+		certPath = defaultKubeletClientCertificateFile
+	}
+	if keyPath == "" {
+		keyPath = defaultKubeletClientKeyFile
+
+	}
+	existingKeyData, err := ioutil.ReadFile(keyPath)
```

Comment:
should we load or generate the key file at this point (something like https://github.com/liggitt/kubernetes/blob/tls-bootstrap/cmd/kubelet/app/bootstrap.go#L65-L82)? it would simplify later methods to always be given key data. also, for out-of-band approval methods, it seems like a generated private key would need to be persisted on the node for another process to be able to answer challenges based on it, which means it has to be written prior to the CSR being submitted.